cluster/addons/dns/kube-dns.yaml.base

```diff
@@ -92,6 +92,16 @@ spec:
         configMap:
           name: kube-dns
           optional: true
+      affinity:
+        podAntiAffinity:
+          requiredDuringSchedulingIgnoredDuringExecution:
```

Comment:
Definitely shouldn't be required. We shouldn't block on scheduling if for some reason (resources or otherwise) we can't meet the affinity requirements. Can you swap this for `preferredDuringSchedulingIgnoredDuringExecution`?

Reply:
Hmm. I'll try to explain my reasoning behind required. I started looking at this since I ran into a situation where all three replicas of kube-dns were running on the same node. That node failed, taking the cluster with it.

From the reliability POV there is never a reason to run multiple replicas of kube-dns on the same node (running one and having the rest pending is the same as running many on one node, as far as reliability is concerned).

The other POV is probably performance, and for that the underlying node doesn't matter much. For a perfect result we'd need "all replicas cannot be on the same node", but I don't think it's possible to specify that..

Reply:
I'll let @bowei and @MrHohn weigh in, but neither POV is worth blocking scheduling IMO. The auto-scaler should pick the right number of replicas. Preferring to schedule them not on the same node makes sense. Refusing to schedule them because there is an existing replica already there doesn't make kube-dns more resilient. `preferredDuringSchedulingIgnoredDuringExecution` will not schedule two kube-dns pods on the same node *unless* there is no other node available to satisfy the affinity rule.

Reply:
In my experience it is quite common for a densely populated cluster to momentarily be in a state where pods can only fit on the "wrong" node. Based on this I'm not sure preferred is a complete fix for this problem.

Reply:
I would prefer to use this patch which is from the previous release cycle. Also, there is a lot of comments regarding why preferred is to be used than required. You will also need to make a corresponding change to kubeadm.

https://github.com/kubernetes/kubernetes/pull/52193

Reply:
@bowei could you please elaborate a bit more why you prefer preferred over required as me https://github.com/kubernetes/kubernetes/pull/57683#issuecomment-356198281 and others https://github.com/kubernetes/kubernetes/pull/57683#issuecomment-356381122 don't see it this way and I would like to understand why you are preferring preferred. The linked PR seems also not to include any details.